File: concat-code.js
#!/usr/bin/env node

/**
 * concat-code.js
 * 
 * This script concatenates all code files from the src directory into a single file.
 * It preserves the original file paths as comments for reference.
 * 
 * Usage: node concat-code.js [output_file]
 */

const fs = require('fs').promises;
const fsSync = require('fs');
const path = require('path');

// Configuration
const srcDir = path.resolve('./src');
const outputFile = process.argv[2] || 'all-code.txt';
const codeExtensions = ['.ts', '.tsx', '.js', '.jsx', '.css', '.html'];

// Stats
let fileCount = 0;

/**
 * Main function using async/await
 */
async function main() {
  console.log(`Starting concatenation process...`);
  console.log(`Source directory: ${srcDir}`);
  console.log(`Output file: ${outputFile}`);
  console.log(`Included extensions: ${codeExtensions.join(', ')}`);
  
  try {
    // Verify source directory exists
    try {
      await fs.access(srcDir);
    } catch (error) {
      console.error(`Error: Source directory ${srcDir} not found.`);
      console.error(`Current working directory: ${process.cwd()}`);
      process.exit(1);
    }
    
    // Create header for output file
    let outputContent = `// All code files from ${srcDir}\n`;
    outputContent += `// Generated on ${new Date().toISOString()}\n`;
    outputContent += `// Extensions included: ${codeExtensions.join(', ')}\n\n`;
    
    // Process files recursively
    outputContent += await walkAndConcatenate(srcDir);
    
    // Check if any files were processed
    if (fileCount === 0) {
      console.log(`No files with extensions ${codeExtensions.join(', ')} found in ${srcDir}.`);
      process.exit(0);
    }
    
    // Write output
    await fs.writeFile(outputFile, outputContent, 'utf8');
    console.log(`\nCompleted! ${fileCount} files concatenated to ${outputFile}`);
  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

/**
 * Recursively walk directory and concatenate file contents
 * @param {string} dirPath - Directory path to process
 * @returns {Promise<string>} - Concatenated content
 */
async function walkAndConcatenate(dirPath) {
  let content = '';
  
  try {
    // Get directory contents
    const items = await fs.readdir(dirPath);
    
    // Process each item
    for (const item of items) {
      const itemPath = path.join(dirPath, item);
      
      try {
        // Get item stats
        const stats = await fs.stat(itemPath);
        
        if (stats.isDirectory()) {
          // Recurse into subdirectories
          content += await walkAndConcatenate(itemPath);
        } else if (stats.isFile()) {
          // Check file extension
          const ext = path.extname(itemPath);
          
          if (codeExtensions.includes(ext)) {
            try {
              // Read file content
              const fileContent = await fs.readFile(itemPath, 'utf8');
              
              // Add file separator and content
              content += `// ===== File: ${itemPath} =====\n\n`;
              content += fileContent;
              content += '\n\n';
              
              // Increase counter and log progress
              fileCount++;
              console.log(`Processed (${fileCount}): ${itemPath}`);
            } catch (readError) {
              console.error(`Error reading file ${itemPath}: ${readError.message}`);
            }
          }
        }
      } catch (itemError) {
        console.error(`Error processing ${itemPath}: ${itemError.message}`);
      }
    }
  } catch (dirError) {
    console.error(`Error reading directory ${dirPath}: ${dirError.message}`);
  }
  
  return content;
}

// Run the main function
main().catch(error => {
  console.error('Unhandled error:', error);
  process.exit(1);
});

File: concat-code.py
#!/usr/bin/env python3

"""
concat-code.py

This script concatenates all code files from the src directory into a single file.
It preserves the original file paths as comments for reference.

Usage: python concat-code.py [output_file]
"""

import os
import sys
import datetime

# Configuration
src_dir = "src"
output_file = sys.argv[1] if len(sys.argv) > 1 else "all-code.txt"
code_extensions = [".ts", ".tsx", ".js", ".jsx", ".css", ".html"]

# Stats
file_count = 0

# Header for output file
header = f"// All code files from {src_dir}\n"
header += f"// Generated on {datetime.datetime.now().isoformat()}\n"
header += f"// Extensions included: {', '.join(code_extensions)}\n\n"

print(f"Concatenating code files from {src_dir} to {output_file}...")

# Function to concatenate files recursively
def concat_files(directory, output_content):
    global file_count
    
    # Walk through all files and directories
    for root, dirs, files in os.walk(directory):
        for file in files:
            # Get full file path
            file_path = os.path.join(root, file)
            
            # Check if file extension is in our list
            _, ext = os.path.splitext(file)
            if ext in code_extensions:
                try:
                    # Try to read file with UTF-8 encoding first
                    try:
                        with open(file_path, 'r', encoding='utf-8') as f:
                            file_content = f.read()
                    except UnicodeDecodeError:
                        # Fall back to latin-1 encoding which can handle any byte sequence
                        with open(file_path, 'r', encoding='latin-1') as f:
                            file_content = f.read()
                            print(f"Note: Used latin-1 encoding for {file_path}")
                    
                    # Add file separator and content
                    output_content += f"\n\n// ===== File: {file_path} =====\n\n"
                    output_content += file_content
                    
                    # Log progress
                    file_count += 1
                    print(f"Processed ({file_count}): {file_path}")
                except Exception as e:
                    print(f"Error reading {file_path}: {str(e)}")
    
    return output_content

try:
    # Check if source directory exists
    if not os.path.isdir(src_dir):
        print(f"Error: Source directory {src_dir} not found.")
        sys.exit(1)
    
    # Concatenate all files
    all_content = concat_files(src_dir, header)
    
    # Check if any files were processed
    if file_count == 0:
        print(f"No files with extensions {', '.join(code_extensions)} found in {src_dir}.")
        sys.exit(0)
    
    # Write output file
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(all_content)
    
    print(f"\nCompleted! {file_count} files concatenated to {output_file}")
except Exception as e:
    print(f"Error: {str(e)}")
    sys.exit(1)